File: lib/services/api_service.dart
// lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../config/environment.dart';

class ApiService {
  // Use environment-based URL configuration
  static String get baseUrl => Environment.apiUrl;
  static String? _authToken;

  // Token management
  static Future<void> setAuthToken(String token) async {
    _authToken = token;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('auth_token', token);
  }

  static Future<String?> getAuthToken() async {
    if (_authToken != null) return _authToken;

    final prefs = await SharedPreferences.getInstance();
    _authToken = prefs.getString('auth_token');
    return _authToken;
  }

  static Future<void> clearAuthToken() async {
    _authToken = null;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('auth_token');
  }

  // HTTP Headers
  static Future<Map<String, String>> _getHeaders({bool requireAuth = false}) async {
    final headers = <String, String>{
      'Content-Type': 'application/json; charset=utf-8',
      'Accept': 'application/json',
      'Accept-Charset': 'utf-8',
    };

    if (requireAuth) {
      final token = await getAuthToken();
      if (token != null) {
        headers['Authorization'] = 'Bearer $token';
      }
    }

    return headers;
  }

  // Clean request data to avoid JSON parsing issues
  static Map<String, dynamic> _cleanRequestData(Map<String, dynamic> data) {
    final cleanData = <String, dynamic>{};

    for (final entry in data.entries) {
      final key = entry.key;
      final value = entry.value;

      if (value is String) {
        // Remove or escape problematic characters
        final cleanValue = value
            .replaceAll('\n', '\\n')
            .replaceAll('\r', '\\r')
            .replaceAll('\t', '\\t')
            .replaceAll('\b', '\\b')
            .replaceAll('\f', '\\f')
            .replaceAll('"', '\\"')
            .trim();
        cleanData[key] = cleanValue;
      } else {
        cleanData[key] = value;
      }
    }

    return cleanData;
  }

  // Generic HTTP methods
  static Future<dynamic> get(String endpoint, {bool requireAuth = false}) async {
    try {
      final fullUrl = '$baseUrl$endpoint';
      print('ApiService GET: $fullUrl');

      final response = await http.get(
        Uri.parse(fullUrl),
        headers: await _getHeaders(requireAuth: requireAuth),
      );

      print('ApiService GET Response: ${response.statusCode}');
      if (response.statusCode != 200) {
        print('ApiService GET Response Body: ${response.body}');
      }

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else if (response.statusCode == 404) {
        throw UnsupportedEndpointException(endpoint, 'Endpoint not found - ${response.body}');
      } else if (response.statusCode == 405) {
        throw UnsupportedEndpointException(endpoint, 'Method not allowed - ${response.body}');
      } else {
        throw ApiException(response.statusCode, response.body);
      }
    } catch (e) {
      print('ApiService GET Exception: $e');
      if (e is UnsupportedEndpointException || e is ApiException) {
        rethrow;
      }
      throw ApiException(0, e.toString());
    }
  }

  static Future<dynamic> post(
      String endpoint,
      Map<String, dynamic> data,
      {bool requireAuth = false}
      ) async {
    try {
      print('Making POST request to: $baseUrl$endpoint');

      // Clean and validate data before encoding
      final cleanData = _cleanRequestData(data);
      final jsonString = json.encode(cleanData);

      print('Request data: $jsonString');
      print('Request data length: ${jsonString.length}');

      final response = await http.post(
        Uri.parse('$baseUrl$endpoint'),
        headers: await _getHeaders(requireAuth: requireAuth),
        body: jsonString,
      ).timeout(
        const Duration(seconds: 30),
        onTimeout: () {
          throw Exception('Request timeout - server took too long to respond');
        },
      );

      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        return json.decode(response.body);
      } else if (response.statusCode == 404) {
        throw UnsupportedEndpointException(endpoint, 'Endpoint not found');
      } else if (response.statusCode == 405) {
        throw UnsupportedEndpointException(endpoint, 'Method not allowed');
      } else {
        throw ApiException(response.statusCode, response.body);
      }
    } catch (e) {
      print('API Error: $e');
      print('API Error type: ${e.runtimeType}');

      // Handle different types of network errors
      if (e.toString().contains('Failed to fetch') ||
          e.toString().contains('XMLHttpRequest') ||
          e.toString().contains('CORS')) {
        print('DEBUG: Network/CORS error detected');
        throw ApiException(0, 'Cannot connect to server. Please check if the backend is running and CORS is configured properly.');
      }

      // Handle timeout errors
      if (e.toString().contains('timeout') || e.toString().contains('TimeoutException')) {
        print('DEBUG: Request timeout detected');
        throw ApiException(0, 'Request timeout. The server is taking too long to respond. Please try again.');
      }

      // Handle JSON parsing errors
      if (e.toString().contains('JSON') || e.toString().contains('parse')) {
        print('DEBUG: JSON parsing error detected');
        throw ApiException(0, 'Server response format error. Please try again or contact support.');
      }

      if (e is UnsupportedEndpointException || e is ApiException) {
        rethrow;
      }
      throw ApiException(0, e.toString());
    }
  }

  static Future<dynamic> put(
      String endpoint,
      Map<String, dynamic> data,
      {bool requireAuth = true}
      ) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl$endpoint'),
        headers: await _getHeaders(requireAuth: requireAuth),
        body: json.encode(data),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        return json.decode(response.body);
      } else if (response.statusCode == 404) {
        throw UnsupportedEndpointException(endpoint, 'Endpoint not found');
      } else if (response.statusCode == 405) {
        throw UnsupportedEndpointException(endpoint, 'Method not allowed');
      } else {
        throw ApiException(response.statusCode, response.body);
      }
    } catch (e) {
      if (e is UnsupportedEndpointException || e is ApiException) {
        rethrow;
      }
      throw ApiException(0, e.toString());
    }
  }

  static Future<dynamic> delete(String endpoint, {bool requireAuth = true}) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl$endpoint'),
        headers: await _getHeaders(requireAuth: requireAuth),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        return response.body.isNotEmpty ? json.decode(response.body) : null;
      } else if (response.statusCode == 404) {
        throw UnsupportedEndpointException(endpoint, 'Endpoint not found');
      } else if (response.statusCode == 405) {
        throw UnsupportedEndpointException(endpoint, 'Method not allowed');
      } else {
        throw ApiException(response.statusCode, response.body);
      }
    } catch (e) {
      if (e is UnsupportedEndpointException || e is ApiException) {
        rethrow;
      }
      throw ApiException(0, e.toString());
    }
  }
}

class ApiException implements Exception {
  final int statusCode;
  final String message;

  ApiException(this.statusCode, this.message);

  @override
  String toString() => 'ApiException: $statusCode - $message';
}

class UnsupportedEndpointException implements Exception {
  final String endpoint;
  final String message;

  UnsupportedEndpointException(this.endpoint, this.message);

  @override
  String toString() => 'UnsupportedEndpointException: $endpoint - $message';
}

File: src/services/apiService.js
// API Service for backend communication
class ApiService {
  static baseUrl = 'https://insidelab.up.railway.app/api/v1';
  static authToken = null;
  static refreshToken = null;
  static isRefreshing = false;
  static refreshSubscribers = [];

  // Token management
  static setAuthToken(token) {
    this.authToken = token;
    localStorage.setItem('auth_token', token);
  }

  static getAuthToken() {
    if (this.authToken) return this.authToken;
    this.authToken = localStorage.getItem('auth_token');
    return this.authToken;
  }

  static setRefreshToken(token) {
    this.refreshToken = token;
    localStorage.setItem('refresh_token', token);
  }

  static getRefreshToken() {
    if (this.refreshToken) return this.refreshToken;
    this.refreshToken = localStorage.getItem('refresh_token');
    return this.refreshToken;
  }

  static clearAuthToken() {
    this.authToken = null;
    this.refreshToken = null;
    localStorage.removeItem('auth_token');
    localStorage.removeItem('refresh_token');
  }

  // HTTP headers
  static getHeaders(requireAuth = false) {
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    if (requireAuth) {
      const token = this.getAuthToken();
      if (token) {
        headers['Authorization'] = `Bearer ${token}`;
      }
    }

    return headers;
  }

  // Refresh access token using refresh token
  static async refreshAccessToken() {
    const refreshToken = this.getRefreshToken();
    if (!refreshToken) {
      throw new Error('No refresh token available');
    }

    try {
      const response = await fetch(`${this.baseUrl}/auth/token/refresh/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ refresh: refreshToken }),
      });

      if (!response.ok) {
        throw new Error('Token refresh failed');
      }

      const data = await response.json();
      if (data.access) {
        this.setAuthToken(data.access);
        return data.access;
      }

      throw new Error('No access token in refresh response');
    } catch (error) {
      console.error('Error refreshing token:', error);
      this.clearAuthToken();
      throw error;
    }
  }

  // Handle token refresh with queue
  static async handleTokenRefresh() {
    if (!this.isRefreshing) {
      this.isRefreshing = true;

      try {
        const newToken = await this.refreshAccessToken();
        this.refreshSubscribers.forEach(callback => callback(newToken));
        this.refreshSubscribers = [];
        return newToken;
      } catch (error) {
        this.refreshSubscribers.forEach(callback => callback(null));
        this.refreshSubscribers = [];
        throw error;
      } finally {
        this.isRefreshing = false;
      }
    }

    // If already refreshing, wait for the result
    return new Promise((resolve) => {
      this.refreshSubscribers.push((token) => {
        resolve(token);
      });
    });
  }

  // Generic fetch with token refresh retry
  static async fetchWithAuth(url, options, requireAuth = false) {
    try {
      const response = await fetch(url, options);

      // If 401 and we have a refresh token, try to refresh
      if (response.status === 401 && requireAuth && this.getRefreshToken()) {
        console.log('Token expired, attempting refresh...');

        try {
          const newToken = await this.handleTokenRefresh();

          if (newToken) {
            // Retry the original request with new token
            options.headers['Authorization'] = `Bearer ${newToken}`;
            const retryResponse = await fetch(url, options);
            return retryResponse;
          }
        } catch (refreshError) {
          console.error('Token refresh failed:', refreshError);
          // Let the original 401 response be handled
        }
      }

      return response;
    } catch (error) {
      throw error;
    }
  }

  // Generic HTTP methods
  static async get(endpoint, requireAuth = false) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method: 'GET',
      headers: this.getHeaders(requireAuth),
    };

    const response = await this.fetchWithAuth(url, options, requireAuth);

    if (!response.ok) {
      const errorData = await response.text();
      throw new ApiException(response.status, errorData);
    }

    return response.json();
  }

  static async post(endpoint, data, requireAuth = false) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method: 'POST',
      headers: this.getHeaders(requireAuth),
      body: JSON.stringify(data),
    };

    const response = await this.fetchWithAuth(url, options, requireAuth);

    if (!response.ok) {
      const errorData = await response.text();
      throw new ApiException(response.status, errorData);
    }

    return response.json();
  }

  static async put(endpoint, data, requireAuth = false) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method: 'PUT',
      headers: this.getHeaders(requireAuth),
      body: JSON.stringify(data),
    };

    const response = await this.fetchWithAuth(url, options, requireAuth);

    if (!response.ok) {
      const errorData = await response.text();
      throw new ApiException(response.status, errorData);
    }

    return response.json();
  }

  static async patch(endpoint, data, requireAuth = false) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method: 'PATCH',
      headers: this.getHeaders(requireAuth),
      body: JSON.stringify(data),
    };

    const response = await this.fetchWithAuth(url, options, requireAuth);

    if (!response.ok) {
      const errorData = await response.text();
      throw new ApiException(response.status, errorData);
    }

    return response.json();
  }

  static async delete(endpoint, requireAuth = false) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method: 'DELETE',
      headers: this.getHeaders(requireAuth),
    };

    const response = await this.fetchWithAuth(url, options, requireAuth);

    if (!response.ok) {
      const errorData = await response.text();
      throw new ApiException(response.status, errorData);
    }

    // DELETE may return empty response
    const text = await response.text();
    return text ? JSON.parse(text) : {};
  }

  // Lab Interest API methods
  static async addLabInterest(labId) {
    try {
      const response = await this.post('/auth/lab-interests/', { lab: labId }, true);
      return response;
    } catch (error) {
      console.error('Error adding lab interest:', error);
      throw error;
    }
  }

  static async removeLabInterest(labId) {
    try {
      const response = await this.post('/auth/lab-interests/remove_interest/', { lab: labId }, true);
      return response;
    } catch (error) {
      console.error('Error removing lab interest:', error);
      throw error;
    }
  }

  static async getLabInterests() {
    try {
      const response = await this.get('/auth/lab-interests/', true);
      return response;
    } catch (error) {
      console.error('Error fetching lab interests:', error);
      throw error;
    }
  }

  // Email and Username validation methods
  static async checkEmailAvailability(email) {
    try {
      const response = await this.post('/auth/check-email/', { email }, false);
      return response;
    } catch (error) {
      console.error('Error checking email availability:', error);
      throw error;
    }
  }

  static async checkUsernameAvailability(username) {
    try {
      const response = await this.post('/auth/check-username/', { username }, false);
      return response;
    } catch (error) {
      console.error('Error checking username availability:', error);
      throw error;
    }
  }
}

// Custom exception for API errors
class ApiException extends Error {
  constructor(statusCode, message) {
    super(message);
    this.statusCode = statusCode;
    this.name = 'ApiException';
  }
}

export { ApiService, ApiException };